clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -4139,6 +4139,56 @@ static bool isCheapEnoughToEvaluateUnconditionally(const Expr *E,
   // exist in the source-level program.
 }
 
+static Value *insertAddressSpaceCast(Value *V, unsigned NewAS) {
+  llvm::PointerType *VTy = cast<llvm::PointerType>(V->getType());
+  if (VTy->getAddressSpace() == NewAS) {
+    return V;
+  }
+  llvm::PointerType *VTyNewAS =
+      llvm::PointerType::get(VTy->getElementType(), NewAS);
+
+  if (auto *Constant = dyn_cast<llvm::Constant>(V)) {
+    return llvm::ConstantExpr::getAddrSpaceCast(Constant, VTyNewAS);
+  }
+
+  llvm::Instruction *NewV =
+      new llvm::AddrSpaceCastInst(V, VTyNewAS, V->getName() + ".ascast.rhs");
```

Comment:
You generate new address space casts not only for "rhs" values but for "lhs" values too. I suggest remove ".rhs" from new instruction name.